device_backends/LogicalNameMapping/src/LNMBitRangeAccessPlugin.cc

```diff
@@ -0,0 +1,232 @@
+    }
+
+    /********************************************************************************************************************/
+
+    void doPreRead(TransferType type) override {
+      _lock.lock();
+
+      _target->preRead(type);
+    }
+
+    /********************************************************************************************************************/
+
+    void doPostRead(TransferType type, bool hasNewData) override {
+      auto unlock = cppext::finally([this] { this->_lock.unlock(); });
+      _target->postRead(type, hasNewData);
+      if(!hasNewData) return;
+
+      if constexpr(std::is_same<UserType, ChimeraTK::Void>::value) {
+        // This should never be reached, the decorate function should make sure that we only have integral types here
+        assert(false);
+      }
+      else if constexpr(std::is_same<UserType, std::string>::value) {
+        // This should never be reached, the decorate function should make sure that we only have integral types here
+        assert(false);
+      }
+      else if constexpr(std::is_same<UserType, ChimeraTK::Boolean>::value) {
+        assert(false);
+      }
+      else {
+        uint64_t v{};
+        auto targetValue = _target->accessData(0);
+        memcpy(std::addressof(v), std::addressof(targetValue), sizeof(TargetType));
+        v = (v & _mask) >> _shift;
+        memcpy(buffer_2D[0].data(), &v, sizeof(UserType));
+      }
+
+      this->_versionNumber = {};
+      this->_dataValidity = _target->dataValidity();
+    }
+
+    /********************************************************************************************************************/
+
+    void doPreWrite(TransferType type, VersionNumber /*versionNumber*/) override {
+      _lock.lock();
+
+      if(!_target->isWriteable()) {
+        throw ChimeraTK::logic_error("This register with BitRange plugin is not writeable.");
+      }
+
+      uint64_t value{};
+      if constexpr(std::is_same<UserType, ChimeraTK::Void>::value) {
+        // This should never be reached, the decorate function should make sure that we only have integral types here
+        assert(false);
+      }
+      else if constexpr(std::is_same<UserType, std::string>::value) {
+        // This should never be reached, the decorate function should make sure that we only have integral types here
+        assert(false);
+      }
+      else if constexpr(std::is_same<UserType, ChimeraTK::Boolean>::value) {
+        assert(false);
+      }
+      else {
+        memcpy(&value, buffer_2D[0].data(), sizeof(UserType));
+        value <<= _shift;
+      }
+
+      // When in a transfer group, only the first accessor to write to the _target can call read() in its preWrite()
+      // Otherwise it will overwrite the
+      if(_target->isReadable() && (!TransferElement::_isInTransferGroup || _lock.useCount() == 1)) {
+        _target->read();
+      }
+
+      _target->accessData(0) &= ~_mask;
+      _target->accessData(0) |= value;
```

Comment:
I think the code would be easier to understand if the shift operation (line 138) would be done here at the r.h.s. of this assignment operation.